Path resolution docs (current state only) v2 (#11442)
ethereum/solidity: docs/path-resolution.rst

   you provide input in JSON format, containing, among other things, the content of all your source
   files:

   .. code-block:: json

       {
           "language": "Solidity",
           "sources": {
               "contract.sol": {
                   "content": "import \"./util.sol\";\ncontract C {}"
               },
               "util.sol": {
                   "content": "library Util {}"
               },
               "/usr/local/dapp-bin/token.sol": {
                   "content": "contract Token {}"
               }
           },
           "settings": {"outputSelection": {"*": { "*": ["metadata", "evm.bytecode"]}}}
       }

   The ``sources`` dictionary becomes the initial content of the virtual filesystem and its keys
   are used as source unit names.

#. **Standard JSON (via import callback)**

   With Standard JSON it is also possible to tell the compiler to use the import callback to obtain
   the source code:

   .. code-block:: json

       {
           "language": "Solidity",
           "sources": {
               "/usr/local/dapp-bin/token.sol": {
                   "urls": [

>                        "/projects/mytoken.sol",

Reviewer: It's not `file://*`?

Author: Both work there. These entries are not really URLs despite the name. They're source unit names passed to the import callback. The Host Filesystem Loader (i.e. the default callback) strips `file://` prefixes. Other callback may be have differently. For example Remix could actually handle URLs.

                       "https://example.com/projects/mytoken.sol"
                   ]
               }
           },
           "settings": {"outputSelection": {"*": { "*": ["metadata", "evm.bytecode"]}}}
       }

>    If an import callback is available, the compiler will pass it the source unit names specified in

Reviewer: the compiler will pass ~it~ the source unit names specified

Author: `it` here means the callback. Maybe it would be clearer as
```
If an import callback is available, the compiler will give it the source unit names specified in
```
?

Author: Replaced with `give`.

source unit name.

.. note::

    A source unit name is just an identifier and even if its value happens to look like a path, it
    is not subject to the normalization rules you would typically expect from filesystem paths.
    Any ``/./`` or ``/../`` seguments or sequences of multiple slashes remain a part of it.
    When the source is provided via Standard JSON interface it is entirely possible to associate
    different content with source unit names that would refer to the same file on disk.

When the source is not available in the virtual filesystem, the compiler passes the source unit name
to the import callback.
The Host Filesystem Loader will attempt to use it as a path and look up the file on disk.
At this point the platform-specific normalization rules kick in and ``/project/lib/math.sol`` and
``/project/lib/../lib///math.sol`` may actually result in the same file being loaded.
Note, however, that the compiler will still see them as separate source units that just happen to
have identical content.

.. index:: ! relative import, ! import; relative
.. _relative-imports:

Relative Imports
----------------

> An import starting with ``./`` or ``../`` is a *relative import*.

Reviewer: What about imports like e.g. `bla/../../`?

Reviewer: Ah ok, I guess these will be treated as direct imports, right?

Author: Yeah. It would be a relative path in the filesystem but not in compiler's VFS. There's a note below about just that.

2. Then the prefix is prepended to the normalized import path.
   If the prefix is non-empty, a single slash is inserted between it and the import path.

In the above the removal of the last segment is performed as follows:

1. Everything past the last slash is removed (i.e. ``a/b//c.sol`` becomes ``a/b//``).
2. All trailing slashes are removed (i.e. ``a/b//`` becomes ``a/b``).

The normalization rules are the same as for UNIX paths, namely:

- All the internal ``./`` segments are removed.
- Every internal ``../``  segment backtracks one level up in the hierarchy.
- Multiple slashes are squashed into a single one.

Note that normalization is performed only on the import path.
The source unit name of the importing module that is used for the prefix remains unnormalized.
This ensures that the ``protocol://`` part does not turn into ``protocol:/`` if the importing file
is identified with a URL.

If your import paths are already normalized, you can expect the above algorithm to produce very
intiuitive results.
Here are some examples of what you can expect if they are not:

.. code-block:: solidity
    :caption: lib/src/../contract.sol

>     import "./util/./util.sol";         // source unit name: lib/src/../util/util.sol

Reviewer: Shouldn't this become `lib/src/util/util.sol`?

Reviewer: Ah ok, so its relative to `lib/src/../contract.sol`.

Author: Yeah. `/../` stays in there because the source unit name of the importing module does not get normalized.

filesystem but force you to keep all files within a single directory and its subdirectories.
When using external libraries it is often desirable to keep their files in a separate location.
To help with that, the compiler provides another mechanism: import remapping.

Remapping allows you to have the compiler replace import path prefixes with something else.
For example you can set up a remapping so that everything imported from the virtual directory
``github.com/ethereum/dapp-bin/library`` would actually receive source unit names starting with
``dapp-bin/library``.
By setting base path to ``/project`` you could then have the compiler find them in
``/project/dapp-bin/library``

The remappings can depend on a context, which allows you to configure packages to import,
e.g. different versions of a library of the same name.

.. warning::

    Information about used remappings is stored in contract metadata so modifying them will result
    in a slightly different bytecode.
    This means that if you move your project files to different locations and use remappings to
    avoid having to adjust the source code, your project will compile but will no longer produce the

>     exact same bytecode as without the remappings.

Reviewer: The bytecode is different, because it's embedded metadata-hash is different, right?

Author: Exactly. I have modified the text now to say that explicitly.

Import remappings have the form of ``context:prefix=target``.
All files in or below the ``context`` directory that import a file that starts with ``prefix`` are
redirected by replacing ``prefix`` with ``target``.
For example, if you clone ``github.com/ethereum/dapp-bin/`` locally to ``/project/dapp-bin``,
you can use the following in your source file:

::

    import "github.com/ethereum/dapp-bin/library/iterable_mapping.sol" as it_mapping;

Then run the compiler:

.. code-block:: bash

    solc github.com/ethereum/dapp-bin/=dapp-bin/ --base-path /project source.sol

As a more complex example, suppose you rely on a module that uses an old version of dapp-bin that
you checked out to ``/project/dapp-bin_old``, then you can run:

.. code-block:: bash

    solc module1:github.com/ethereum/dapp-bin/=dapp-bin/ \
         module2:github.com/ethereum/dapp-bin/=dapp-bin_old/ \
         --base-path /project \
         source.sol

> This means that all imports in ``module2`` point to the old version but imports in ``module1``

Reviewer: I do not really understand `context` here. I understand the `prefix=target` part, but what does `context` really mean? How can it be used?

Reviewer: After reading the rest of this file, I understood how this works. But I didn't understood it when I read about this the first time here. Maybe it makes sense to put here also an example, e.g.:
```
Parts of file and directory names can match as well.
/newProject/con:/new=old will match /newProject/contract.sol 
and remap it to oldProject/contract.sol.
```
But I'm still not sure how this mechanism could be useful.

Author: I have rewritten that section and also added this, which hopefully makes the purpose of `context` clearer:

```
You can limit the scope of a remapping by specifying a *context*.
This allows creating remappings that apply only to imports located in a specific library or a specific file.
Without a context a remapping is applied to every matching import in all the files in the virtual
filesystem.
```

> Maybe it makes sense to put here also an example, e.g.:

I don't really want to draw attention to this because for me it's more like an unintended consequence than a feature. If we were to simplify remappings I would actually make them match only the whole segments ending with `/`. I think the current behavior is completely unexpected for most people (it was for me).

point to the new version.

Here are the detailed rules governing the behaviour of remappings:

#. **Remappings only affect the translation between import paths and source unit names.**

   Source unit names added via other means cannot be remapped.
   For example the paths you specify on the command-line and the ones in ``sources.urls`` in
   Standard JSON are not affected.

    .. code-block:: bash

>         solc /project=/contracts /project/contract.sol # source unit name: /project/contract.sol

Reviewer: Where does the data come from? Is it loading the data from `/project/contract.sol` or is the file loaded from `/contract/contract.sol`?

Author: `/project/contract.sol`. Remappings only affect source unit names created from imports. Paths on the command-line are not remapped at all.

I have now added a sentence below that example to make it clear.

#. **Context and prefix must match source unit names, not import paths.**

   - This means that you cannot remap ``./`` or ``../`` directly since they are replaced during
     the translation to source unit name but you can remap the part of the name they are replaced
     with:

     .. code-block:: bash

>          solc ./=a /project=b /project/contract.sol

Reviewer: Similar as above. I think it would be also nice to add the source unit name here too, but I'm really not sure from where the data will be loaded. I understand that the source unit name for this file will result in `/project/contract.sol`, but was the content loaded from a file located at `b/contract.sol`?

Reviewer: I think adding the source unit name here would be helpful :) I don't know anymore whether the resulting source unit name will be `/project/contract.sol` or `b/contract.sol`.

Author: It's `/project/contract.sol` because paths given on the command-line are not remapped.

Added a comment to clarify that.

     When using the JavaScript interface you can even use URLs and abstract identifiers if
     your callback can handle them.

   - Remapping happens after relative imports have already been resolved into source unit names.
     This means that targets starting with ``./`` and ``../`` have no special meaning and are
     relative to the base path rather than to the location of the source file.

   - Remapping targets are not normalized so ``@root=./a/b//`` will remap ``@root/contract.sol``
     to ``./a/b//contract.sol`` and not ``a/b/contract.sol``.

   - If the target does not end with a slash, the compiler will not add one automatically:

     .. code-block:: bash

         solc /project/=/contracts /project/contract.sol

     .. code-block:: solidity
         :caption: /project/contract.sol

         import "/project/util.sol" as util; // source unit name: /contractsutil.sol

#. **Context and prefix are patterns and matches must be exact.**

   - ``a//b=c`` will not match ``a/b``.
   - source unit names are not normalized so ``a/b=c`` will not match ``a//b`` either.
   - Parts of file and directory names can match as well.

>      ``/newProject/con:/new=old`` will match ``/newProject/contract.sol`` and remap it to

Reviewer: Oha! That looks really crazy! I really wonder whether this mechanism is really used.

Author: I doubt it is. But people using it should be aware of it and know to always include the final `/`. Otherwise the results can be pretty confusing :)

                       "https://example.com/projects/mytoken.sol"
                   ]
               }
           },
           "settings": {"outputSelection": {"*": { "*": ["metadata", "evm.bytecode"]}}}
       }

   If an import callback is available, the compiler will give it the source unit names specified in
   ``urls`` one by one, until one is loaded successfully or the end of the list is reached.

   The source unit names are determined the same way as when using ``content`` - they are keys of
   the ``sources`` dictionary and the content of ``urls`` does not affect them in any way.

   .. index:: standard input, stdin, <stdin>

#. **Standard input**

   On the command line it is also possible to provide the source is by sending it to compiler's
   standard input:

   .. code-block:: bash

       echo 'import "./util.sol"; contract C {}' | solc -

   ``-`` used as one of the arguments instructs the compiler to place the content of the standard
   input in the virtual filesystem under a special source unit name: ``<stdin>``.

Once the VFS is initialized, additional files can still be added to it only through the import
callback.

.. index:: ! import; path

Imports
=======

> The import statement specifies an *import path*, which, after being lightly processed, becomes a

Reviewer: Maybe we could also add here `import "<import-path>"`, to ensure that everyone understands the definition of an import path.

Author: Done. I added an example with two imports and described which part is the import path and how it relates to the source unit name.

    import "@openzeppelin/address.sol";     // source unit name: @openzeppelin/address.sol
    import "https://example.com/token.sol"; // source unit name: https://example.com/token.sol

After applying any :ref:`import remappings <import-remapping>` the import path simply becomes the
source unit name.

.. note::

    A source unit name is just an identifier and even if its value happens to look like a path, it
    is not subject to the normalization rules you would typically expect from filesystem paths.
    Any ``/./`` or ``/../`` seguments or sequences of multiple slashes remain a part of it.
    When the source is provided via Standard JSON interface it is entirely possible to associate
    different content with source unit names that would refer to the same file on disk.

When the source is not available in the virtual filesystem, the compiler passes the source unit name
to the import callback.
The Host Filesystem Loader will attempt to use it as a path and look up the file on disk.
At this point the platform-specific normalization rules kick in and ``/project/lib/math.sol`` and
``/project/lib/../lib///math.sol`` may actually result in the same file being loaded.

> Note, however, that the compiler will still see them as separate source units that just happen to

Reviewer: Maybe we could put this also in a `.. note::` section?

Author: I tried to change it but left it as is after all. This sentence is strongly connected to this paragraph and I think that putting in a note (or even just in a separate paragraph) makes it read weird.

Reviewer 2: It's rather confusing because the two import paths somehow appear out of nowhere and it's not clear what they refer to.

Author: ok, now I get what the problem is. I have reworded that paragraph and changed how the examples are introduced so that now it makes sense as an actual note.

    *direct import with a relative path* (e.g. ``import "util.sol"``).
    It is a relative import only if it starts with ``./`` or ``../``.
    Otherwise it is a direct import.

The compiler computes a source unit name from the import path in the following way:

1. First a prefix is computed

    - Prefix is initialized with the source unit name of the importing source unit.
    - The last segment is removed from the prefix.
    - Then, for every leading ``../`` segment in the normalized import path the last segment is repeatedly removed from the prefix.

2. Then the prefix is prepended to the normalized import path.
   If the prefix is non-empty, a single slash is inserted between it and the import path.

In the above the removal of the last segment is performed as follows:

1. Everything past the last slash is removed (i.e. ``a/b//c.sol`` becomes ``a/b//``).
2. All trailing slashes are removed (i.e. ``a/b//`` becomes ``a/b``).

The normalization rules are the same as for UNIX paths, namely:

- All the internal ``./`` segments are removed.
- Every internal ``../``  segment backtracks one level up in the hierarchy.
- Multiple slashes are squashed into a single one.

> Note that normalization is performed only on the import path.

Reviewer: Maybe we could emphasize that this only applies for relative import paths. E.g. `Note that normalization is performed only on *relative* import paths.`

Author: My emphasis here was actually on the fact that it's performed on import path (i.e. only on the thing you see in the import statement) as opposed to the complete source unit name resulting from the import. I think that mentioning that it's relative here would detract from that. And the whole section is about relative imports anyway.

The source unit name of the importing module that is used for the prefix remains unnormalized.
This ensures that the ``protocol://`` part does not turn into ``protocol:/`` if the importing file

> is identified with a URL.

Reviewer: Maybe here we could say that URL's are always direct imports, and thats why they don't get normalized.

Author: I could mention that in the section about URLs but I think that it's a bit off topic at this particular place.

The situation here is that the source unit name of the **importing module** is a URL. Maybe I'll try to make that clearer.

Author: I did not add it after all. I had this in my original PR but removed it exactly because Chris wanted me to only describe the rules. It's going to be included in the blog post though.

Reviewer: Ah great! I need to check the blog post then too. Is there already a PR for that - did not check yet?

Author: Not yet. I wanted to get the base docs out first.

If your import paths are already normalized, you can expect the above algorithm to produce very
intiuitive results.
Here are some examples of what you can expect if they are not:

.. code-block:: solidity
    :caption: lib/src/../contract.sol

    import "./util/./util.sol";         // source unit name: lib/src/../util/util.sol
    import "./util//util.sol";          // source unit name: lib/src/../util/util.sol
    import "../util/../array/util.sol"; // source unit name: lib/src/array/util.sol

>     import "../.././../util.sol";       // source unit name: util.sol

Reviewer: When I read this I was thinking what would be the source unit name of `import "../../../../../../util.sol";` I guess it will also be `util.sol`. Maybe we could also add this.

Author: ok. I'll add this particular one. One example won't hurt.

I don't want to add too many examples overall though because I cut them out from the original #11263 on purpose to keep the docs concise. They will be included in the blog post instead.

Author: Added.

#. **Context and prefix are patterns and matches must be exact.**

   - ``a//b=c`` will not match ``a/b``.
   - source unit names are not normalized so ``a/b=c`` will not match ``a//b`` either.
   - Parts of file and directory names can match as well.
     ``/newProject/con:/new=old`` will match ``/newProject/contract.sol`` and remap it to
     ``oldProject/contract.sol``.

#. **At most one remapping is applied to a single import.**

   - If multiple remappings match the same source unit name, the one with the longest matching
     prefix is chosen.
   - If prefixes are identical, the one specified last wins.
   - Remappings do not work on other remappings. For example ``a=b b=c c=d`` will not result in ``a``
     being remapped to ``d``.

#. **Prefix cannot be empty but context and target are optional.**

   If ``target`` is omitted, it defaults to the value of the ``prefix``.

.. note::

    ``solc`` only allows you to include files from certain directories.
    They have to be in the directory (or subdirectory) of one of the explicitly specified source
    files or in the directory (or subdirectory) of a remapping target.

>     If you want to allow direct absolute includes, add the remapping ``/=/``.

Reviewer: I would say we should state here, that it is not recommended to do.

Author: I decided to remove this part after all. It was a part I just copied from the original docs. It's actually pretty weird because you can get the same effect with `--allow-paths /` and using a remapping looks like a roundabout way to do it. Maybe it predates `--allow-paths`?

In any case, we don't really want anyone to do this :)

.. _path-resolution:

***************

> Path Resolution

Reviewer 2: Maybe "Import Path Resolution"?

Author: Well, it's not just about imports (also for CLI arguments and paths used in JSON) so I think "Path Resolution" is more accurate but I can change it if you think it gives people reading the docs a better idea of that this section is about.

Reviewer 2: ok!

Author: Changed the header. I left the file name and reference as is.

***************

In order to be able to support reproducible builds on all platforms, the Solidity compiler has to
abstract away the details of the filesystem where source files are stored.
Paths used in imports must work the same way everywhere while the command-line interface must be
able to work with platform-specific paths to provide good user experience.
This section aims to explain in detail how Solidity reconciles these requirements.

.. index:: ! virtual filesystem, ! VFS, ! source unit name
.. _virtual-filesystem:

Virtual Filesystem
==================

The compiler maintains an internal database (*virtual filesystem* or *VFS* for short) where each
source unit is assigned a unique *source unit name* which is an opaque and unstructured identifier.
When you use the :ref:`import statement <import>`, you specify an *import path* that references a
source unit name.

.. index:: ! import callback, ! Host Filesystem Loader
.. _import-callback:

Import Callback
---------------

The VFS is initially populated only with files the compiler has received as input.

> Additional files can be loaded during compilation using an *import callback*.

Reviewer 2: ```suggestion
Additional files can be loaded during compilation using an *import callback*, which is
different depending on the type of compiler you use (see below).
```

If the compiler does not find any source unit name matching the import path in the VFS, it invokes
the callback, which is responsible for obtaining the source code to be placed under that name.

> An import callback is free to interpret source unit names in an arbitrary way, not just a paths.

Reviewer 2: ```suggestion
An import callback is free to interpret source unit names in an arbitrary way, not just as paths.
```

If there is no callback available when one is needed or if it fails to locate the source code,
compilation fails.

The command-line compiler provides the *Host Filesystem Loader* - a rudimentary callback
that interprets a source unit name as a path in the local filesystem.
The `JavaScript interface <https://github.com/ethereum/solc-js>`_ does not provide any by default,
but one can be provided by the user.
This mechanism can be used to obtain source code from locations other then the local filesystem
(which may not even be accessible, e.g. when the compiler is running in a browser).
For example the `Remix IDE <https://remix.ethereum.org/>`_ provides a versatile callback that
lets you `import files from HTTP, IPFS and Swarm URLs or refer directly to packages in NPM registry
<https://remix-ide.readthedocs.io/en/latest/import.html>`_.

.. note::

    Host Filesystem Loader's file lookup is platform-dependent.
    For example backslashes in source unit name can be interpreted as directory separators or not
    and the lookup can be case-sensitive or not, depending on the underlying platform.

>     For portability it is recommended to avoid using import paths that will work correctly only

Reviewer 2: Shouldn't we suggest to always use forward slashes (which will also work on windows), for example?

Author: I thought that it should be clear combined with the previous paragraph. I can state it more explicitly.

Reviewer 2: It was not clear to me :)

Reviewer 2: And I mean we should really put a clear recommendation here (in its own notice or whatever), not just casually mention it in the text.

Author: > It was not clear to me :)

Fair enough.

> And I mean we should really put a clear recommendation here (in its own notice or whatever), not just casually mention it in the text.

Not sure what you mean. This part is already inside a note. Do you want it to be in a second, separate note? I think it would be a bit weird because both notes would be talking about the same thing - one in general and one about a very specific case.

    with a specific import callback or only on one platform.

.. index:: ! CLI path

> Populating the Virtual Filesystem

Reviewer 2: ```suggestion
Initial Contents of the Virtual Filesystem
```
?

Author: Either is fine in my opinion so I changed it. But what's wrong with it? My version was a bit shorter so more suited for a header.

Reviewer 2: To me, "populating" is the whole process of filling it up, including via the load callback.

Author: ok. Could be interpreted that way.

---------------------------------

The initial content of the VFS depends on how you invoke the compiler:

> #. **CLI**

Reviewer 2: ```suggestion
#. **solc / commandline interface**
```

Author: Actually, this is something that badly needs a name. What do we call the command-line interface excluding `--standard-json`? I need to refer to it pretty often and I started generally referring to it as CLI even though I know it's not precise. I just do not have a better name and I don't think "solc / commandline interface" is all that much better (`--standard-json` is also a part of solc).

"Shell interface"? "non-JSON interface"? "Command mode"?

Reviewer 2: I think command line interface is good, I would just not use the abbreviation and also clarify how to get to it (mention "solc").

Reviewer 2: Defining a name for us does not mean that readers will know what we mean :)

Author: I don't mean just define it and leave it there :) I mean also using it. And of course not just an arbitrary name completely unrelated to the thing. I think that a good name could stick. The problem is just finding one that's actually good :)

   When you compile a file using the command-line interface of the compiler, you provide one or

>    more *CLI paths* to files containing Solidity code:

Reviewer 2: Shouldn't we use the term "commandline argument"?

Reviewer 2: I mean that's the question I'm asking: Is the commandline argument interpreted by the compiler in any way or not?

Author: It is but only in a very minimal way. It goes through Boost's `generic_string()` and as far as I can tell that only normalizes path separators. For example it does not strip or modify Windows drive letter (at least on Linux) or sequences of multiple slashes.

I'll change "CLI path" to just path. Initially I wanted to use this term to differentiate it from paths we get from elsewhere but I see that I ended up using it only in this paragraph so I can drop the term.

   .. code-block:: bash

       solc contract.sol /usr/local/dapp-bin/token.sol

>    The source unit name of a file loaded this way is simply the specified CLI path with

Reviewer 2: Should we briefly mention the word "shell expansion" or would that be too detailed?

Author: Technically it's irrelevant here since you could use the executable without a shell but I guess adding that detail might make it clearer to some users. I'll add it.

               "contract.sol": {
                   "content": "import \"./util.sol\";\ncontract C {}"
               },
               "util.sol": {
                   "content": "library Util {}"
               },
               "/usr/local/dapp-bin/token.sol": {
                   "content": "contract Token {}"
               }
           },
           "settings": {"outputSelection": {"*": { "*": ["metadata", "evm.bytecode"]}}}
       }

   The ``sources`` dictionary becomes the initial content of the virtual filesystem and its keys
   are used as source unit names.

#. **Standard JSON (via import callback)**

   With Standard JSON it is also possible to tell the compiler to use the import callback to obtain
   the source code:

   .. code-block:: json

       {
           "language": "Solidity",
           "sources": {
               "/usr/local/dapp-bin/token.sol": {
                   "urls": [
                       "/projects/mytoken.sol",
                       "https://example.com/projects/mytoken.sol"
                   ]
               }
           },
           "settings": {"outputSelection": {"*": { "*": ["metadata", "evm.bytecode"]}}}
       }

>    If an import callback is available, the compiler will give it the source unit names specified in

Reviewer 2: ```suggestion
   If an import callback is available, the compiler will give it the strings specified in
```

   ``urls`` one by one, until one is loaded successfully or the end of the list is reached.

   The source unit names are determined the same way as when using ``content`` - they are keys of
   the ``sources`` dictionary and the content of ``urls`` does not affect them in any way.

   .. index:: standard input, stdin, <stdin>

#. **Standard input**

>    On the command line it is also possible to provide the source is by sending it to compiler's

Reviewer: On the command line it is also possible to provide the source ~is~ by sending it to compiler's

>    On the command line it is also possible to provide the source is by sending it to compiler's

Reviewer 2: ```suggestion
   On the command line it is also possible to provide the source by sending it to compiler's
```

   standard input:

   .. code-block:: bash

       echo 'import "./util.sol"; contract C {}' | solc -

   ``-`` used as one of the arguments instructs the compiler to place the content of the standard
   input in the virtual filesystem under a special source unit name: ``<stdin>``.

Once the VFS is initialized, additional files can still be added to it only through the import
callback.

.. index:: ! import; path

Imports
=======

> The import statement specifies an *import path*, which, after being lightly processed, becomes a

Reviewer 2: ```suggestion
The import statement specifies an *import path*. This path, when relative (it starts with `./` or `../`), is combined with the path of the importing source unit to form the imported source unit name. If it is not relative, it is taken as the imported source unit name directly.
```

Reviewer 2: Ok, this ignores remappings. But still, I think we should replace "lightly processed" by a short summary of the processing.

Author: I moved the paragraph about two kinds of import above the example and mentioned `./` and `../`. Is that ok?

source unit names they translate to are ``contracts/math/math.sol`` and ``contracts/tokens/token.sol``
respectively.

Based on how the processing is performed, we can divide imports into two categories:

- :ref:`Direct imports <direct-imports>`, where you specify the full source unit name.
- :ref:`Relative imports <relative-imports>`, where you specify a path to be combined with the
  source unit name of the importing file.

.. index:: ! direct import, import; direct
.. _direct-imports:

Direct Imports
--------------

An import that does not start with ``./`` or ``../`` is a *direct import*.

::

    import "/project/lib/util.sol";         // source unit name: /project/lib/util.sol
    import "lib/util.sol";                  // source unit name: lib/util.sol
    import "@openzeppelin/address.sol";     // source unit name: @openzeppelin/address.sol
    import "https://example.com/token.sol"; // source unit name: https://example.com/token.sol

After applying any :ref:`import remappings <import-remapping>` the import path simply becomes the
source unit name.

.. note::

    A source unit name is just an identifier and even if its value happens to look like a path, it

>     is not subject to the normalization rules you would typically expect from filesystem paths.

Reviewer 2: ```suggestion
    is not subject to the normalization rules you would typically expect from shells.
```

``/project/lib/../lib///math.sol`` may actually result in the same file being loaded.
Note, however, that the compiler will still see them as separate source units that just happen to
have identical content.

.. index:: ! relative import, ! import; relative
.. _relative-imports:

Relative Imports
----------------

An import starting with ``./`` or ``../`` is a *relative import*.
Such imports specify a path relative to the source unit name of the importing source unit:

.. code-block:: solidity
    :caption: /project/lib/math.sol

    import "./util.sol" as util;    // source unit name: /project/lib/util.sol
    import "../token.sol" as token; // source unit name: /project/token.sol

.. code-block:: solidity
    :caption: lib/math.sol

    import "./util.sol" as util;    // source unit name: lib/util.sol
    import "../token.sol" as token; // source unit name: token.sol

.. note::

    Note the difference between a *relative import* (e.g. ``import "./util.sol"``) and a
    *direct import with a relative path* (e.g. ``import "util.sol"``).
    It is a relative import only if it starts with ``./`` or ``../``.

>     Otherwise it is a direct import.

Reviewer 2: I'm worried that the term "direct import with a relative path" could be confusing. Maybe we could say that unless the import starts with `./` or `../` it is a direct import and is thus absolute and imports do not have to start with `/` to be absolute or something...

Author: How about this?
```rst
.. note::

    ``import "./util.sol"`` is a relative import but ``import "util.sol"`` is a direct one.
    Relative imports **must** start with ``./`` or ``../``.
    VFS does not have a root directory and absolute paths do not have to start with a slash.
    While both paths would be considered relative in the host filesystem, ``util.sol`` is actually
    absolute in the VFS.
```

Reviewer 2: ("must start" sound like it is an error otherwise)
```
.. note::

    Imports that start with `./` or `../` are relative imports, while all others are direct (absolute) imports.
    Direct imports do to need to start with `/`. This, ``import "./util.sol"`` is a relative import but ``import "util.sol"`` is a direct one.
    While both paths would be considered relative in the host filesystem, ``util.sol`` is actually
    absolute in the VFS.
```
"The VFS does not have a root directory" - of course it has, that's where all the topmost files are. You just don't need to use `/` to reference the root directory.

But yeah, this is probably not too important.

Author: I really, really do not want to use the term *absolute import* :)

Point taken about `must`. Replaced with `always`, which I hope does not have this connotation.

> "The VFS does not have a root directory" - of course it has, that's where all the topmost files are. You just don't need to use `/` to reference the root directory.

I think one could also argue that all the topmost items are just floating there without any directory containing them but that would be splitting hairs and a root with an empty name makes sense too so I just removed that part for simplicity:)

So here's another attempt, combining my and your version:

```rst
.. note::

    Relative imports **always** start with ``./`` or ``../`` so ``import "util.sol"``, unlike
    ``import "./util.sol"``, is a direct import.
    While both paths would be considered relative in the host filesystem, ``util.sol`` is actually
    absolute in the VFS.
```

The compiler computes a source unit name from the import path in the following way:

1. First a prefix is computed

    - Prefix is initialized with the source unit name of the importing source unit.

>     - The last segment is removed from the prefix.

Reviewer 2: ```suggestion
    - The last segment (the file name) is removed from the prefix.
```

Reviewer 2: Also "segment" is introduced here without explanation.

Author: I was sure it was a common term but apparently it's not. The Wikipedia page for [filename](https://en.wikipedia.org/wiki/Filename) does not even mention it.

While segment was not defined, "removal of the last segment" was.

Anyway, added a definition.

>     - Then, for every leading ``../`` segment in the normalized import path the last segment is repeatedly removed from the prefix.

Reviewer 2: This is the first mention of "normalized" - in which way is it normalized?

Author: It's explained below. I think it flows better that way because first you see the outline and only then the details are filled in. I could add "see below" but there would be a lot of that so I thought it was better as it is now.

2. Then the prefix is prepended to the normalized import path.
   If the prefix is non-empty, a single slash is inserted between it and the import path.

In the above the removal of the last segment is performed as follows:

1. Everything past the last slash is removed (i.e. ``a/b//c.sol`` becomes ``a/b//``).
2. All trailing slashes are removed (i.e. ``a/b//`` becomes ``a/b``).

The normalization rules are the same as for UNIX paths, namely:

- All the internal ``./`` segments are removed.
- Every internal ``../``  segment backtracks one level up in the hierarchy.
- Multiple slashes are squashed into a single one.

> Note that normalization is performed only on the import path.

Reviewer 2: ```suggestion
Note that normalization is performed only on the import path, not on the importing source unit name.
```

The source unit name of the importing module that is used for the prefix remains unnormalized.
This ensures that the ``protocol://`` part does not turn into ``protocol:/`` if the importing file
is identified with a URL.

If your import paths are already normalized, you can expect the above algorithm to produce very

> intiuitive results.

Reviewer 2: ```suggestion
intuitive results.
```

Here are some examples of what you can expect if they are not:

.. code-block:: solidity
    :caption: lib/src/../contract.sol

    import "./util/./util.sol";         // source unit name: lib/src/../util/util.sol
    import "./util//util.sol";          // source unit name: lib/src/../util/util.sol
    import "../util/../array/util.sol"; // source unit name: lib/src/array/util.sol
    import "../.././../util.sol";       // source unit name: util.sol
    import "../../.././../util.sol";    // source unit name: util.sol

.. note::

    The use of relative imports containing leading ``../`` segments is not recommended.
    The same effect can be achieved in a more reliable way by using direct imports with
    :ref:`base path <base-path>` and :ref:`import remapping <import-remapping>`.

.. index:: ! base path, --base-path
.. _base-path:

Base Path
=========

> Base path specifies the directory that the Host Filesystem Loader will load files from.

Reviewer 2: ```suggestion
The base path specifies the directory that the Host Filesystem Loader will load files from.
```

It is simply prepended to a source unit name before the filesystem lookup is performed.

> By default base path is empty, which leaves the source unit name unchanged.

Reviewer 2: ```suggestion
By default the base path is empty, which leaves the source unit name unchanged.
```

filesystem.

Import remappings have the form of ``context:prefix=target``:

- ``context`` must match the beginning of the source unit name of the file containing the import.
- ``prefix`` must match the beginning of the source unit name resulting from the import.
- ``target`` is the value the prefix is replaced with.

For example, if you clone https://github.com/ethereum/dapp-bin/ locally to ``/project/dapp-bin``
and run the compiler with:

.. code-block:: bash

    solc github.com/ethereum/dapp-bin/=dapp-bin/ --base-path /project source.sol

you can use the following in your source file:

.. code-block:: solidity

    import "github.com/ethereum/dapp-bin/library/math.sol"; // source unit name: dapp-bin/library/math.sol

The compiler will look for the file in the VFS under ``dapp-bin/library/math.sol``.
If the file is not available there, the source unit name will be passed to the Host Filesystem
Loader, which will then look in ``/project/dapp-bin/library/iterable_mapping.sol``.

.. warning::

    Information about remappings is stored in contract metadata.
    Since the binary produced by the compiler has a hash of the metadata embedded in it, any

>     modification to the remappings will result in a slightly different bytecode.

Reviewer 2: ```suggestion
    modification to the remappings will result in different bytecode.
```

Author: Well, it's only the tiny bit containing the metadata hash that is different :) I did not want to give an impression that it changes the behavior of the contract.

Removed.

   - Parts of file and directory names can match as well.
     ``/newProject/con:/new=old`` will match ``/newProject/contract.sol`` and remap it to
     ``oldProject/contract.sol``.

#. **At most one remapping is applied to a single import.**

   - If multiple remappings match the same source unit name, the one with the longest matching
     prefix is chosen.
   - If prefixes are identical, the one specified last wins.
   - Remappings do not work on other remappings. For example ``a=b b=c c=d`` will not result in ``a``
     being remapped to ``d``.

#. **Prefix cannot be empty but context and target are optional.**

   If ``target`` is omitted, it defaults to the value of the ``prefix``.

.. index:: Remix IDE, file://

Using URLs in imports
=====================

Most URL prefixes such as ``https://`` or ``data://`` have no special meaning in import paths.
The only exception is ``file://`` which is stripped from source unit names by the Host Filesystem
Loader.

When compiling locally you can use import remapping to replace the protocol and domain part with a
local path:

.. code-block:: bash

    solc :https://github.com/ethereum/dapp-bin=/usr/local/dapp-bin contract.sol

> Note the leading ``:``.
> It is necessary when the remapping context is empty.

Reviewer 2: ```suggestion
Note the leading ``:``, which is necessary when the remapping context is empty.
```
